src/main/java/org/spongepowered/common/mixin/core/item/recipe/smelting/MixinFurnaceRecipes.java

```diff
@@ -50,93 +59,70 @@
-     *         normally supported in vanilla.
-     */
-    @Overwrite
-    public void addSmeltingRecipe(ItemStack input, ItemStack stack, float experience) {
-        ItemStackSnapshot exemplaryResult = ItemStackUtil.snapshotOf(stack);
-        ItemStackSnapshot exemplaryIngredient = ItemStackUtil.snapshotOf(input);
-        Predicate<ItemStackSnapshot> ingredientPredicate = new MatchSmeltingVanillaItemStack(exemplaryIngredient);
-
-        this.recipes.add(new SpongeSmeltingRecipe(exemplaryResult, exemplaryIngredient, ingredientPredicate, experience));
-    }
+public abstract class MixinFurnaceRecipes implements SmeltingRecipeRegistry {
+
+    @Shadow @Final private Map<ItemStack, ItemStack> smeltingList;
+    @Shadow @Final private Map<ItemStack, Float> experienceList;
+
+    private final List<SmeltingRecipe> customRecipes = Lists.newArrayList();
+    // No IdentityHashBiMap implementation exists
+    private final Map<SmeltingRecipe, ItemStack> customRecipeToNativeIngredient = new IdentityHashMap<>();
+    private final Map<ItemStack, SmeltingRecipe> nativeIngredientToCustomRecipe = new IdentityHashMap<>();
+
+    @Shadow public abstract ItemStack getSmeltingResult(ItemStack stack);
+    @Shadow public abstract float getSmeltingExperience(ItemStack stack);
 
-    /**
-     * @author Limeth - 17. 1. 2017
-     * @reason Instead of using the {@link Map}s in {@link FurnaceRecipes}, we
-     *         mixin and maintain a list of {@link SmeltingRecipe}s. This is to
-     *         allow recipe resolution via ingredient predicates, which are not
-     *         normally supported in vanilla.
-     */
-    @Overwrite
-    public ItemStack getSmeltingResult(ItemStack stack) {
+    @Inject(method = "getSmeltingResult", at = @At("RETURN"), cancellable = true)
+    public void onGetSmeltingResult(ItemStack stack, CallbackInfoReturnable<ItemStack> cir) {
         ItemStackSnapshot ingredient = ItemStackUtil.snapshotOf(stack);
-        Optional<SmeltingResult> result = getResult(ingredient);
+        Optional<SmeltingResult> result = getCustomResult(ingredient);
 
-        //noinspection OptionalIsPresent
         if (result.isPresent()) {
-            return ItemStackUtil.fromSnapshotToNative(result.get().getResult());
-        } else {
-            return ItemStack.EMPTY;
+            ItemStack nativeResult = ItemStackUtil.fromSnapshotToNative(result.get().getResult());
+
+            cir.setReturnValue(nativeResult);
         }
     }
 
-    /**
-     * @author Limeth - 17. 1. 2017
-     * @reason Instead of using the {@link Map}s in {@link FurnaceRecipes}, we
-     *         mixin and maintain a list of {@link SmeltingRecipe}s. This is to
-     *         allow recipe resolution via ingredient predicates, which are not
-     *         normally supported in vanilla.
-     */
-    @Overwrite
-    public float getSmeltingExperience(ItemStack stack) {
+    @Inject(method = "getSmeltingExperience", at = @At("RETURN"))
```

Comment:
It's using `setReturnValue` so yes, it needs to be cancellable or it will throw an exception if that code path is followed.